src/requirements/README.md

```diff
@@ -1,44 +1,103 @@
+  with additional list of fulfillable courses attached to each requirement. The file
+  [`filtered-all-courses.ts`](./``filtered-all-courses.ts) simply reexported a statically-typed
+  version of it.
+
+### Requirements Sources
+
+The metadata of all requirements (name, source, minimum count, etc) is stored in
+[`source-requirements-json.ts`](./source-requirements-json.ts). It's type is specified by the
+`RequirementsJson` type in [`types.ts`](./types.ts). The meanings of the field has been documented
+as JSDoc directly in type definitions.
+
+The code for precomputing a list of fulfillable courses are in the [`checkers`](./checkers) folder.
+They are organized by colleges and majors. All of those checkers are aggregated together in the
+[`checkers/all-requirements.checker.ts`](./checkers/all-requirements-checkers.ts).
+
+## Architecture
+
+### Pre-computation phase
+
+To reiterate, this phase computes a list of satisfying courses for each requirement.
+
+1. Setup:
+   a. Each college or major requirement in
+   [`source-requirements-json.ts`](./source-requirements-json.ts) has a field called `checkerName`.
+   b. Each checker in
+   [`checkers/all-requirements.checker.ts`](./checkers/all-requirements-checkers.ts) is assigned
+   a checker name as its key.
+2. During pre-computation, we use the `checkerName` value in `source-requirements-json.ts` to find
+   a correct checker, and then run the checker. From that, we get a list of courses.
+3. We attach a list of courses to the requirement metadata in [`generator.ts`](./generator.ts).
+4. Now we created [`decorated-requirements.json`](./decorated-requirements.json).
+   Pre-computation is done.
+
+Each checker is either:
+
+- a TypeScript function `(course: Course) => boolean` that directly tells whether the course
+  satisfies the requirement that this checker correspond to.
+- or a list of such functions, where each function checks whether the course satisfies one of the
+  sub-requirements that this checker correspond to.
+
+This is how you should use your checkers:
+
+- Case 1: the requirement contains no sub-requirement. Use option 1. Example: CS major practicum.
+- Case 2: the requirement contains sub-requirements. Use option 2. Example: CS Core requirement.
+
+### Frontend Computation Phase
+
+#### University Requirements
+
+Since too many courses can satisfy these requirements, the satisfying course list is not
+pre-computed to reduce the data sent from server to client.
+
+It's computed on the client side directly in `computeUniversityRequirementFulfillments` function
+inside [`reqs-functions.ts`](./reqs-functions.ts).
+
+#### College or Major Requirements
+
+1. We first naively compute whether a sub-requirement is fulfilled by checking whether any course in
+   user's list is in the pre-computed satisfying course list.
+2. Then, we run a deduplication pass to remove all courses that have been double-counted that
```

Comment:
Is this where we would check for cross-listed classes fulfilling a single requirement? (like ENGRD 2110 and CS 2110)? Or should all versions of the same class be added in the checker?

Reply:
Currently there is no way. We can add another pass to the pre-computation phase that adds in all cross-listed course. We also need to add another deduplication phase that merges cross listed courses together on client side.